Fuse dict keys into a DICT_KEYS guard if we are guarding on all of them.

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118003
* __->__ #118096
* #117625
* #118098
* #117983
* #117982

We group guards on keys that are "easily printable" and we
transform them into just one `DICT_KEYS`.

It feels a bit silly to first split the guards in the `builder` and then
regroup them, but this allows for a number of things:
1. Correctness. `DICT_KEYS` was trying to do too much before this set of
   PRs. Now `DICT_KEYS` will be reserved for things that are easy to
   print. If they are not then we fallback to the slower but correct
   getitem method. Even more so, if the builder places several guards on
   find which ones can be elided by the `DICT_KEYS` transformation,
   while the others will be kept there for correctness.
2. Can be generalised to lazy keys. We can generalize `DICT_KEYS` so that it
   just checks that a subset of keys exist within the dict. We just have
   this info at guard creation, not at guard instantiation.

For now we just perform this optimisation for CONSTANT_MATCH, but we
plan to generalize this to other types in the future, once we have lazy
key support.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

lezcano committed:
Fuse dict keys into a DICT_KEYS guard if we are guarding on all of them.

We group guards on keys that are "easily printable" and we
transform them into just one `DICT_KEYS`.

It feels a bit silly to first split the guards in the `builder` and then
regroup them, but this allows for a number of things:
1. Correctness. `DICT_KEYS` was trying to do too much before this set of
   PRs. Now `DICT_KEYS` will be reserved for things that are easy to
   print. If they are not then we fallback to the slower but correct
   getitem method. Even more so, if the builder places several guards on
   find which ones can be elided by the `DICT_KEYS` transformation,
   while the others will be kept there for correctness.
2. Can be applied to lazy keys. We can generalize `DICT_KEYS` so that it
   just checks that a subset of keys exist within the dict. We just have
   this info at guard creation, not at guard instantiation.

For now we just perform this optimisation for CONSTANT_MATCH, but we
plan to generalize this to other types in the future, once we have lazy
key support.

[ghstack-poisoned]

peterbell10 commented on torch/_dynamo/guards.py:
```diff
@@ -1018,6 +1021,32 @@ def cleanup_builder(weak_b):
+    @staticmethod
+    def fuse_dict_guards(builder, guards):
+        # Rather than creating n contant guards, each of them costing O(n)
+        # (ConstDictKeySource.reconstruct is O(n), we just place a guard
```
```suggestion
        # (ConstDictKeySource.reconstruct is O(n)), we just place a guard
```

peterbell10 commented on torch/_dynamo/guards.py:
```diff
@@ -1018,6 +1021,32 @@ def cleanup_builder(weak_b):
+        add_guards = set()
+        for (guard, l, elem_guards) in dicts.values():
+            # TODO This condition can be relaxed generalizing DICT_KEYS
+            if len(elem_guards) == l:
```
NIT: `l` (ell) is easily confused with `1` (one). Maybe call this `n_elems`?

peterbell10 reviewed (commented):
Needs some tests but otherwise LGTM.

lezcano committed:
Update on "Fuse dict keys into a DICT_KEYS guard if we are guarding on all of them."


We group guards on keys that are "easily printable" and we
transform them into just one `DICT_KEYS`.

It feels a bit silly to first split the guards in the `builder` and then
regroup them, but this allows for a number of things:
1. Correctness. `DICT_KEYS` was trying to do too much before this set of
   PRs. Now `DICT_KEYS` will be reserved for things that are easy to
   print. If they are not then we fallback to the slower but correct
   getitem method. Even more so, if the builder places several guards on
   find which ones can be elided by the `DICT_KEYS` transformation,
   while the others will be kept there for correctness.
2. Can be generalised to lazy keys. We can generalize `DICT_KEYS` so that it
   just checks that a subset of keys exist within the dict. We just have
   this info at guard creation, not at guard instantiation.

For now we just perform this optimisation for CONSTANT_MATCH, but we
plan to generalize this to other types in the future, once we have lazy
key support.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Fuse dict keys into a DICT_KEYS guard if we are guarding on all of them."


We group guards on keys that are "easily printable" and we
transform them into just one `DICT_KEYS`.

It feels a bit silly to first split the guards in the `builder` and then
regroup them, but this allows for a number of things:
1. Correctness. `DICT_KEYS` was trying to do too much before this set of
   PRs. Now `DICT_KEYS` will be reserved for things that are easy to
   print. If they are not then we fallback to the slower but correct
   getitem method. Even more so, if the builder places several guards on
   find which ones can be elided by the `DICT_KEYS` transformation,
   while the others will be kept there for correctness.
2. Can be generalised to lazy keys. We can generalize `DICT_KEYS` so that it
   just checks that a subset of keys exist within the dict. We just have
   this info at guard creation, not at guard instantiation.

For now we just perform this optimisation for CONSTANT_MATCH, but we
plan to generalize this to other types in the future, once we have lazy
key support.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel reviewed (changes requested):
Add a test

lezcano committed:
Update on "Fuse dict keys into a DICT_KEYS guard if we are guarding on all of them."


We group guards on keys that are "easily printable" and we
transform them into just one `DICT_KEYS`.

It feels a bit silly to first split the guards in the `builder` and then
regroup them, but this allows for a number of things:
1. Correctness. `DICT_KEYS` was trying to do too much before this set of
   PRs. Now `DICT_KEYS` will be reserved for things that are easy to
   print. If they are not then we fallback to the slower but correct
   getitem method. Even more so, if the builder places several guards on
   find which ones can be elided by the `DICT_KEYS` transformation,
   while the others will be kept there for correctness.
2. Can be generalised to lazy keys. We can generalize `DICT_KEYS` so that it
   just checks that a subset of keys exist within the dict. We just have
   this info at guard creation, not at guard instantiation.

For now we just perform this optimisation for CONSTANT_MATCH, but we
plan to generalize this to other types in the future, once we have lazy
key support.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Fuse dict keys into a DICT_KEYS guard if we are guarding on all of them."


We group guards on keys that are "easily printable" and we
transform them into just one `DICT_KEYS`.

It feels a bit silly to first split the guards in the `builder` and then
regroup them, but this allows for a number of things:
1. Correctness. `DICT_KEYS` was trying to do too much before this set of
   PRs. Now `DICT_KEYS` will be reserved for things that are easy to
   print. If they are not then we fallback to the slower but correct
   getitem method. Even more so, if the builder places several guards on
   find which ones can be elided by the `DICT_KEYS` transformation,
   while the others will be kept there for correctness.
2. Can be generalised to lazy keys. We can generalize `DICT_KEYS` so that it
   just checks that a subset of keys exist within the dict. We just have
   this info at guard creation, not at guard instantiation.

For now we just perform this optimisation for CONSTANT_MATCH, but we
plan to generalize this to other types in the future, once we have lazy
key support.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano commented:
Superseded by https://github.com/pytorch/pytorch/pull/118194
